src/Parsers/ExpressionElementParsers.cpp

```diff
@@ -504,6 +516,51 @@ bool ParserTableFunctionView::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
     return true;
 }
 
+bool ParserFilterClause::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+{
+    assert(node);
+    ASTFunction & function = dynamic_cast<ASTFunction &>(*node);
+
+    ParserToken parser_openging_bracket(TokenType::OpeningRoundBracket);
+    if (!parser_openging_bracket.ignore(pos, expected))
+    {
+        return false;
+    }
+
+    ParserKeyword parser_where("WHERE");
+    if (!parser_where.ignore(pos, expected))
+    {
+        return false;
+    }
+    ParserExpressionList parser_condition(false);
+    ASTPtr condition;
+    if (!parser_condition.parse(pos, condition, expected))
+    {
+        return false;
+    }
+
+    ParserToken parser_closing_bracket(TokenType::ClosingRoundBracket);
+    if (!parser_closing_bracket.ignore(pos, expected))
+    {
+        return false;
+    }
+
+    if (function.name.find("If") != String::npos)
+    {
+        throw Exception(
+            ErrorCodes::ILLEGAL_AGGREGATION,
+            "Filter clause provided for an aggregating function (" + function.name + ") already containing If suffix");
+    }
+    if (condition->children.empty())
+    {
+        throw Exception(ErrorCodes::SYNTAX_ERROR, "Empty condition for WHERE");
+    }
+
+    function.name += "If";
```

Comment:
Is to possible to have not aggregate function here, that doesn't support `If` modifier?

Reply:
It should be handled, maybe not in parser but on later stages
```
:) SELECT identity(number) FILTER(WHERE number > 22) FROM numbers(100) ;

SELECT identityIf(number, number > 22)
FROM numbers(100)

Query id: 7001a006-0879-4206-a151-0ae842500997


0 rows in set. Elapsed: 0.017 sec.

Received exception from server (version 21.9.1):
Code: 46. DB::Exception: Received from localhost:9000. DB::Exception: Unknown function identityIf. Maybe you meant: ['identity']: While processing identityIf(number, number > 22). (UNKNOWN_FUNCTION)

------ 

:) SELECT throw(number == 1) FILTER(WHERE number > 22) FROM numbers(100) ;

SELECT throwIf(number = 1, number > 22)
FROM numbers(100)

Query id: 8c13959e-8b82-40a8-b155-516075b2de4f


0 rows in set. Elapsed: 0.003 sec.

Received exception from server (version 21.9.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type UInt8 of argument of function throwIf: While processing throwIf(number = 1, number > 22). (ILLEGAL_TYPE_OF_ARGUMENT)
```

Reply:
> It should be handled, maybe not in parser but on later stages

I would say that CH should not happily accept the wrong query and produce the result (which was the case with sumIf + filter) but +- any error is enough for such weird cases (fooIf doesn't exist or resulting set of args is wrong).